File: lib/LocalModules/DevOther/src/main/java/java/dev/engine/barcode/BarCodeConfig.java
package java.dev.engine.barcode;

import com.google.zxing.DecodeHintType;
import com.google.zxing.EncodeHintType;
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel;

import java.util.EnumMap;
import java.util.Map;

import dev.engine.barcode.IBarCodeEngine;
import dev.utils.DevFinal;

/**
 * detail: BarCode Config
 * @author Ttt
 */
public class BarCodeConfig
        extends IBarCodeEngine.EngineConfig {

    // 编码 ( 生成 ) 配置
    private final Map<EncodeHintType, Object> encodeHints = new EnumMap<>(EncodeHintType.class);
    // 解码 ( 解析 ) 配置
    private final Map<DecodeHintType, Object> decodeHints = new EnumMap<>(DecodeHintType.class);

    // ===========
    // = get/set =
    // ===========

    /**
     * 获取编码 ( 生成 ) 配置
     * @return 编码 ( 生成 ) 配置
     */
    public Map<EncodeHintType, Object> getEncodeHints() {
        return encodeHints;
    }

    /**
     * 获取解码 ( 解析 ) 配置
     * @return 解码 ( 解析 ) 配置
     */
    public Map<DecodeHintType, Object> getDecodeHints() {
        return decodeHints;
    }

    // ==========
    // = 快捷方法 =
    // ==========

    /**
     * 设置默认编码 ( 生成 ) 配置
     * @return BarCode Config
     */
    public BarCodeConfig defaultEncode() {
        // 编码类型
        encodeHints.put(EncodeHintType.CHARACTER_SET, DevFinal.ENCODE.UTF_8);
        // 指定纠错等级, 纠错级别 ( L 7%、M 15%、Q 25%、H 30% )
        encodeHints.put(EncodeHintType.ERROR_CORRECTION, ErrorCorrectionLevel.H);
        // 设置二维码边的空度, 非负数
        encodeHints.put(EncodeHintType.MARGIN, 0);
        return this;
    }

    /**
     * 设置编码 ( 生成 ) 配置
     * @return BarCode Config
     */
    public BarCodeConfig putEncodeHints(Map<EncodeHintType, Object> hints) {
        if (hints != null) encodeHints.putAll(hints);
        return this;
    }

    /**
     * 设置解码 ( 解析 ) 配置
     * @return BarCode Config
     */
    public BarCodeConfig putDecodeHints(Map<DecodeHintType, Object> hints) {
        if (hints != null) decodeHints.putAll(hints);
        return this;
    }
}

File: lib/DevEngine/src/main/java/dev/engine/core/barcode/BarCodeConfig.kt
package dev.engine.core.barcode

import com.google.zxing.DecodeHintType
import com.google.zxing.EncodeHintType
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import dev.engine.barcode.IBarCodeEngine
import dev.utils.DevFinal
import java.util.*
import kotlin.jvm.JvmField

/**
 * detail: BarCode Config
 * @author Ttt
 */
open class BarCodeConfig : IBarCodeEngine.EngineConfig {

    // 编码 ( 生成 ) 配置（m 前缀避免与 [getEncodeHints] JVM 签名冲突）
    @JvmField
    protected val mEncodeHints: MutableMap<EncodeHintType, Any> = EnumMap(
        EncodeHintType::class.java
    )

    // 解码 ( 解析 ) 配置
    @JvmField
    protected val mDecodeHints: MutableMap<DecodeHintType, Any> = EnumMap(
        DecodeHintType::class.java
    )

    // ===========
    // = get/set =
    // ===========

    /**
     * 获取编码 ( 生成 ) 配置
     * @return 编码 ( 生成 ) 配置
     */
    open fun getEncodeHints(): MutableMap<EncodeHintType, Any> {
        return mEncodeHints
    }

    /**
     * 获取解码 ( 解析 ) 配置
     * @return 解码 ( 解析 ) 配置
     */
    open fun getDecodeHints(): MutableMap<DecodeHintType, Any> {
        return mDecodeHints
    }

    // ==========
    // = 快捷方法 =
    // ==========

    /**
     * 设置默认编码 ( 生成 ) 配置
     * @return BarCode Config
     */
    open fun defaultEncode(): BarCodeConfig {
        // 编码类型
        mEncodeHints[EncodeHintType.CHARACTER_SET] = DevFinal.ENCODE.UTF_8
        // 指定纠错等级, 纠错级别 ( L 7%、M 15%、Q 25%、H 30% )
        mEncodeHints[EncodeHintType.ERROR_CORRECTION] = ErrorCorrectionLevel.H
        // 设置二维码边的空度, 非负数
        mEncodeHints[EncodeHintType.MARGIN] = 0
        return this
    }

    /**
     * 设置编码 ( 生成 ) 配置
     * @return BarCode Config
     */
    open fun putEncodeHints(hints: Map<EncodeHintType, Any>?): BarCodeConfig {
        hints?.let { mEncodeHints.putAll(it) }
        return this
    }

    /**
     * 设置解码 ( 解析 ) 配置
     * @return BarCode Config
     */
    open fun putDecodeHints(hints: Map<DecodeHintType, Any>?): BarCodeConfig {
        hints?.let { mDecodeHints.putAll(it) }
        return this
    }
}